### PR TITLE
Flag persistence generators

### PR DESCRIPTION
**Pull request by ulupo**
This PR add the possibility to retrieve representative simplicies.

* [x] : Compute and retrieve finite representative for dimensions > 0
* [x] : Compute and retrieve essential representative for dimensions > 0
* [x] : Compute and retrieve finite representative for dimension == 0
* [x] : Compute and retrieve essential representative for dimension == 0
* [x] : Add support in the bindings for retrieving representative simplicies
* [x] : Add support in Python to retrieve and output representative simplicies
    * [x] : Document the output regarding representative simplicies
    * [x] : Throw exception if `collapser` and `representative simplicies` are enable at the same time
* [x] : Update documentation about representative simplicies
* [x] : Update example basic tutorial to show example of this feature

### CODE REVIEW

**Comment by ulupo** [comment]
> (i for column and j for row)
We are playing with fire here :) I suggest we stick to `i` indexing rows and `j` columns as this is virtually universal in mathematics and Python.

**Comment by ulupo** [comment]
During work on this PR #29, @MonkeyBreaker and I realized that the current core logic in the matrix reduction algorithm makes it so that we break the following promise to the user: that passing a dense matrix with a few `np.inf` will lead to exactly the same output as passing a sparse version where the `np.inf` entries are just missing entries.  This was never noticed at the level of barcodes because a "finite" bar with death equal to infinity looks just the same as an infinite bar with the same birth.  But flag generators don't lie: we now end up telling the user there are never any generators for essential classes if the matrix is dense, no matter what the matrix is or represents.

Because changing this requires potentially delicate changes to the reduction routine, we decided to discuss what to do about this separately, and open a separate PR if necessary after this one is merged.

**Comment by ulupo** [comment]
In view of #36 , we need to update the imports in the two notebooks here.

**Comment by MonkeyBreaker** [comment]
> In view of #36 , we need to update the imports in the two notebooks here.

Before doing the merge, we will need to rebase to `main` and then we will be able to merge, otherwise as you point out, we will have conflicts

File: gph/src/ripser.h [hunk 1]
```diff
@@ -1163,6 +1229,52 @@ class ripser
...
+                            edge = new_cand;
+                        }
+                    }
+                }
+            }
+        }
+
+        return edge;
+    }
+
```
[the remark this answers is not in the document]
Comment (MonkeyBreaker) [reply gph/src/ripser.h:1296]: This was done on commit 0abb23b
I think we can resolve the issue, thank you Umbe for the suggestion of using a diamong input

**Comment by MonkeyBreaker** [comment]
> Another note: res['gens'][1] and res['gens'][3] should always have length maxdim. Currently, they can be empty lists if there is no content. Instead, we should always return a list of length maxdim with empty ndarrays if necessary in some (or even all) positions.

It seems weird to me, the C++ core has the same logic as for barcodes. The only difference is that you have `maxdim-1` as size of res['gens'][1] and res['gens'][3] , because dimension 0 is not included.

I need to verify if this is not the case

**Comment by ulupo** [comment]
> The only difference is that you have maxdim as size of res['gens'][1] and res['gens'][3] , because dimension 0 is not included.

I don't think this can be the case if `maxdim` means the same as in Python. Example: if `maxdim=1` (compute up to and including dimension 1) then the list has length 1, not 1-1=0.

File: gph/src/ripser.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ulupo) [reply gph/src/ripser.h:1296]: Thanks for implementing it as a test!!

**Comment by MonkeyBreaker** [comment]
> I don't think this can be the case if maxdim means the same as in Python. Example: if maxdim=1 (compute up to and including dimension 1) then the list has length 1, not 1-1=0.

Well when you compute `maxdim=1` you have dimension 0 and dimension 1 => 2-1 = 1, no ?

**Comment by ulupo** [comment]
Oh I'm sorry @MonkeyBreaker , I thought you said `maxdim - 1` (because it used to say that in the docstrings), but you didn't! We said the same thing.

**Comment by MonkeyBreaker** [comment]
> Oh I'm sorry @MonkeyBreaker , I thought you said maxdim - 1 (because it used to say that in the docstrings), but you didn't! We said the same thing, i.e. that res['gens'][1] and res['gens'][3] should have length maxdim.

Does it work as expected then ? Do we have an issue in the documentation ? (In the documentation we are right in that we return `maxdim` for [1] and [3].

**Comment by ulupo** [comment]
> Does it work as expected then ? 

I think I was able to produce examples which returned empty lists when `maxdim=1`, but can't remember how.  Maybe `square_dm`?

**Comment by MonkeyBreaker** [comment]
>  Maybe square_dm?

Just tried with `maxdim=1` works as expected

File: examples/flag_persistence_generators.ipynb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7e4bf17b-f6b4-444c-a344-38ff096c7f4a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Giotto-ph\n",
+    "from gph.python import ripser_parallel\n",
```
Comment (ulupo) [inline examples/flag_persistence_generators.ipynb:33]: The import here was not updated correctly, I wonder if something went wrong in resolving conflicts @MonkeyBreaker ?

Comment (ulupo) [inline examples/flag_persistence_generators.ipynb:33]:
```suggestion
    "from gph import ripser_parallel\n",
```

Comment (ulupo) [reply examples/flag_persistence_generators.ipynb]: I fixed it, and I suspect nothing went wrong. Merging!

Comment (MonkeyBreaker) [reply examples/flag_persistence_generators.ipynb]: :angry: I failed the merge, sorry for that

File: gph/python/test/test_ripser.py [hunk 3]
```diff
@@ -189,3 +190,123 @@ def test_multithread_consistent(dm):
...
+        # Verify birth of dim-0 essential representative
+        # The birth is located on the diagonal
+        assert np.isclose(barcode[0], X[rp, rp])
+
+
+def test_gens_order_vertices_higher_dimension():
+    """This test verifies that function get_youngest_edge_simplex
+       returns the correct vertices. We are interested that in the computation
+       it will return the edge with the youngest vertices a the biggest diameter
+    """
```
Comment (ulupo) [inline gph/python/test/test_ripser.py:296]:
```suggestion
    """This test verifies that function `get_youngest_edge_simplex` returns the
    correct vertices. It should return the edge with the largest diameter in
    the simplex and, if several are present with the same diameter, the oldest
    one in the reverse colexicographic order used to build the simplexwise
    refinement of the Vietoris-Rips filtration."""
```

**Pull request merged**